shared/src/http_transport.rs

```diff
@@ -0,0 +1,187 @@
+        let message = format!("failed to send request: {}", err);
+        tracing::debug!("[id:{}] {}", id, message);
+        Web3Error::Transport(message)
+    })?;
+    let status = response.status();
+    let text = response.text().await.map_err(|err| {
+        let message = format!("failed to get response body: {}", err);
+        tracing::debug!("[id:{}] {}", id, message);
+        Web3Error::Transport(message)
+    })?;
+    // Log the raw text before decoding to get more information on responses that aren't valid
+    // json. Debug encoding so we don't get control characters like newlines in the output.
+    tracing::debug!("[id:{}] received response: {:?}", id, text.trim());
+    if !status.is_success() {
+        return Err(Web3Error::Transport(format!(
+            "response status code is not success: {}",
+            status
+        )));
+    }
+    jsonrpc_core::serde_from_str(&text).map_err(Into::into)
+}
+
+type RpcResult = Result<Value, Web3Error>;
+
+impl Transport for HttpTransport {
+    type Out = BoxFuture<'static, RpcResult>;
+
+    fn prepare(&self, method: &str, params: Vec<Value>) -> (RequestId, Call) {
+        let id = self.next_id();
+        let request = helpers::build_request(id, method, params);
+        (id, request)
+    }
+
+    fn send(&self, id: RequestId, call: Call) -> Self::Out {
+        let (client, url) = self.new_request();
+        async move {
+            let output = execute_rpc(client, url, id, &Request::Single(call)).await?;
+            helpers::to_result_from_output(output)
+        }
+        .boxed()
+    }
+}
+
+impl BatchTransport for HttpTransport {
+    type Batch = BoxFuture<'static, Result<Vec<RpcResult>, Web3Error>>;
+
+    fn send_batch<T>(&self, requests: T) -> Self::Batch
+    where
+        T: IntoIterator<Item = (RequestId, Call)>,
+    {
+        // Batch calls don't need an id but it helps associate the response log to the request log.
+        let id = self.next_id();
+        let (client, url) = self.new_request();
+        let (ids, calls): (Vec<_>, Vec<_>) = requests.into_iter().unzip();
+        async move {
+            let outputs = execute_rpc(client, url, id, &Request::Batch(calls)).await?;
+            handle_batch_response(&ids, outputs)
+        }
+        .boxed()
+    }
+}
+
+fn handle_batch_response(
+    ids: &[RequestId],
+    outputs: Vec<Output>,
+) -> Result<Vec<RpcResult>, Web3Error> {
+    let mut outputs = outputs
+        .into_iter()
+        .map(|output| {
+            Ok((
+                id_of_output(&output)?,
+                helpers::to_result_from_output(output),
+            ))
+        })
+        .collect::<Result<HashMap<_, _>, Web3Error>>()?;
+    ids.iter()
+        .map(|id| {
+            outputs.remove(id).ok_or_else(|| {
+                Web3Error::InvalidResponse(format!("batch response is missing id {}", id))
+            })
+        })
+        .collect()
```

Comment:
Given the spec:
> A Response object SHOULD exist for each Request object, except that there SHOULD NOT be any Response objects for notifications.

Should we also check that there are no remaining responses in the `outputs` after all the requests have been matched?

Reply:
Will add this.